internals/overlord/pairingstate/manager.go

```diff
@@ -0,0 +1,277 @@
+
+// pairingstate manages client server pairing.
+package pairingstate
+
+import (
+	"context"
+	"crypto/x509"
+	"errors"
+	"fmt"
+	"strconv"
+	"strings"
+	"sync"
+	"time"
+
+	"github.com/canonical/pebble/internals/overlord/state"
+	"github.com/canonical/pebble/internals/plan"
+)
+
+// timeAfterFunc allows faking time.AfterFunc.
+var timeAfterFunc = func(d time.Duration, f func()) Timer {
+	return time.AfterFunc(d, f)
+}
+
+// Timer is used so we can supply a fake timer during testing.
+type Timer interface {
+	Stop() bool
+}
+
+// Mode controls the pairing policy of the pairing manager.
+type Mode string
+
+const (
+	// ModeUnset is the same as ModeDisabled, but this value prevents the
+	// plan from marshalling the Mode explicitly.
+	ModeUnset Mode = ""
+	// ModeDisabled means no pairing is possible
+	ModeDisabled Mode = "disabled"
+	// ModeSingle means only a single client can pair.
+	ModeSingle Mode = "single"
+	// ModeMultiple means multiple clients can pair.
+	ModeMultiple Mode = "multiple"
+)
+
+type PairingConfig struct {
+	Mode Mode `yaml:"mode,omitempty"`
+}
+
+func (c *PairingConfig) Validate() error {
+	switch c.Mode {
+	case ModeUnset, ModeDisabled, ModeSingle, ModeMultiple:
+	default:
+		return fmt.Errorf("cannot support pairing mode %q: unknown mode", c.Mode)
+	}
+	return nil
+}
+
+// Implements the optional Zeroer interface as used by the YAML library
+// for deciding when to marshal a section or not.
+func (c *PairingConfig) IsZero() bool {
+	return c.Mode == ModeUnset
+}
+
+func (c *PairingConfig) Combine(other *PairingConfig) {
+	if other.Mode != ModeUnset {
+		c.Mode = other.Mode
+	}
+}
+
+type PairingManager struct {
+	state  *state.State
+	mu     sync.Mutex
+	config *PairingConfig
+	open   bool
+	// timer controls the duration of the pairing window.
+	timer Timer
+	// cancelTimerFunc cancels a pairing window if enabled.
+	cancelTimerFunc context.CancelFunc
+}
+
+func NewManager(state *state.State) *PairingManager {
+	m := &PairingManager{
+		state: state,
+		config: &PairingConfig{
+			Mode: ModeDisabled,
+		},
+	}
+	return m
+}
+
+// PlanChanged informs the pairing manager that the plan has been updated.
+func (m *PairingManager) PlanChanged(update *plan.Plan) {
+	m.mu.Lock()
+	defer m.mu.Unlock()
+
+	newConfig := update.Sections[PairingField].(*PairingConfig)
+
+	// If the mode changed, force the pairing window to be reopened
+	// taking the new config into account.
+	if m.config.Mode != newConfig.Mode {
+		m.open = false
+		m.stopTimer()
+	}
+	m.config = newConfig
+}
+
+// Ensure implements StateManager.Ensure.
+func (m *PairingManager) Ensure() error {
+	return nil
+}
+
+// PairMTLS adds a client identity with admin permissions to the identity
+// subsystem. A pairing request always closes the pairing window.
+func (m *PairingManager) PairMTLS(clientCert *x509.Certificate) error {
+	m.mu.Lock()
+	defer m.mu.Unlock()
+
+	if !m.open {
+		return errors.New("cannot pair client: pairing is not open")
```

Comment:
Shouldn't this simply be

```suggestion
		return errors.New("pairing is not open")
```

and when calling from other places:

```go
if err := m.PairMTLS(clientCert); err != nil {
    return fmt.Errorf("cannot pair client: %w", err)
}

```